### PR TITLE
Stop non-finite steps before optimizer updates

### PR DESCRIPTION
**Pull request by taking-lying-flat**
## Summary

- check loss finiteness on every training step, independent of metric logging frequency
- combine loss and gradient-norm finiteness flags across the full distributed world
- stop before optimizer and learning-rate scheduler updates when any rank reports a non-finite value
- handle both regular tensors and DTensor loss values

## Root cause

The existing non-finite loss check lived in the metrics logging path. Non-logging steps skipped it entirely, and logging steps reached the check only after `optimizer.step()` and `lr_scheduler.step()`. A NaN or Inf could therefore corrupt model parameters and optimizer state before training stopped.

## Impact

Every gradient-accumulation loss and the clipped gradient norm are now checked through device-side flags. A world-wide reduction makes all ranks take the same failure path before the optimizer update, including pipeline stages that do not own the loss.

- Tradeoff: `_assert_async` avoids a per-step host sync; on CPU it raises a catchable `RuntimeError`, while on CUDA a failed assertion invalidates the process.

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -944,6 +951,19 @@ def train_step(
                 pp_mesh=parallel_dims.get_optional_mesh("pp"),
                 ep_enabled=parallel_dims.ep_enabled,
             )
+            finite_flags = torch.stack(
+                [loss_is_finite, torch.isfinite(grad_norm).all()]
+            ).to(dtype=torch.int32)
+            torch.distributed.all_reduce(
+                finite_flags,
+                op=torch.distributed.ReduceOp.MIN,
+                group=parallel_dims.world_mesh.get_group(),
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: question 1: should be able to use `loss_mesh`, instead of world_mesh, because e.g. across all TP ranks the value should be the same

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -944,6 +951,19 @@ def train_step(
                 pp_mesh=parallel_dims.get_optional_mesh("pp"),
                 ep_enabled=parallel_dims.ep_enabled,
             )
+            finite_flags = torch.stack(
+                [loss_is_finite, torch.isfinite(grad_norm).all()]
+            ).to(dtype=torch.int32)
+            torch.distributed.all_reduce(
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: question 2: do we need to all-reduce before erroring out, or put in another way, do we need all ranks to error out together? I think
- even if some rank failed, it's going to be detected and all ranks would stop
- the introduction of this small collective seems not a big deal, but I'd like to understand the tradeoff

**Comment by tianyu-l** [comment]
@claude help review this PR. In particular, comment on my review feedback.

File: torchtitan/trainer.py [hunk 3]
```diff
@@ -944,6 +951,19 @@ def train_step(
...
+                finite_flags,
+                op=torch.distributed.ReduceOp.MIN,
+                group=parallel_dims.world_mesh.get_group(),
+            )
+            if not finite_flags.all().item():
+                raise RuntimeError(
+                    "Loss or gradient norm is not finite on at least one rank at "
+                    f"step {self.step}. Stopping training before the optimizer update."
+                )
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: yeah as claude said, this is very bad as it creates device-host sync -- I think that's the reason we didn't do this before

File: torchtitan/trainer.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/trainer.py]: we probably should let PP last rank always broadcast loss to other ranks @sanketpurandare

**Comment by tianyu-l** [comment]
@claude review again

**Comment by tianyu-l** [comment]
@taking-lying-flat 

> The remaining must-fix is the reverted unit test, which now encodes the old logging-gated behavior and will fail / misrepresent the change.

Please fix.

> I'd also add one line to the PR body on the _assert_async vs catchable-RuntimeError tradeoff.

Let's add this note into the trainer code.

**Review by tianyu-l — APPROVED** [review]
